src/main/java/org/folio/rest/impl/LoginAPI.java

```diff
@@ -288,184 +268,255 @@ public void postAuthnLogin(String userAgent, String xForwardedFor,
+          }
+          userVerified.onComplete(verifyResult -> loginAfterUserVerified(verifyResult, entity, newTenantId, okapiURL,
+              requestToken, userAgent, xForwardedFor, okapiHeaders, asyncResultHandler, vertxContext));
+        });
+    } catch(Exception e) {
+      logger.debug("Error running on verticle for postAuthnLogin: " + e.getLocalizedMessage());
+      asyncResultHandler.handle(Future.succeededFuture(PostAuthnLoginResponse.respond500WithTextPlain(INTERNAL_ERROR)));
+    }
+  }
 
-        } else {
-          //User's okay, let's try to login
-          try {
-            JsonObject userObject = verifyResult.result();
-            if (!userObject.containsKey("id")) {
-              logger.error("No 'id' key in returned user object");
-              asyncResultHandler.handle(Future.succeededFuture(
-                  PostAuthnLoginResponse.respond500WithTextPlain(
-                      "No user id could be found")));
+  private Future<String> handleCrossTenantLogin(LoginCredentials credentials, String tenantId,
+                                                Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler) {
+    if (credentials.getTenantId() == null || credentials.getTenantId().equals(okapiHeaders.get(XOkapiHeaders.TENANT)))
+      return Future.succeededFuture(tenantId);
+    Promise<String> promise = Promise.promise();
+    userService.getUserTenants(tenantId, credentials.getUsername(), credentials.getUserId(), credentials.getTenantId(),
+      LoginConfigUtils.encodeJsonHeaders(okapiHeaders), ar -> {
+        if (ar.failed()) {
+          promise.fail(ar.cause());
+          return;
+        }
+        try {
+          List<UserTenant> userTenants = ar.result().stream()
+            .map(obj -> ((JsonObject)obj).mapTo(UserTenant.class))
+            .collect(toList());
+          if (userTenants.isEmpty()) {
+            if (StringUtils.isNotBlank(credentials.getTenantId())) {
+              logger.info(MISSING_USER_TENANT_ASSOCIATION_LOG, credentials.getUsername(), credentials.getUserId(),
+                  credentials.getTenantId());
+              asyncResultHandler.handle(Future.succeededFuture(PostAuthnLoginResponse.respond422WithApplicationJson(
+                  LoginAPI.getErrors(MISSING_USER_TENANT_ASSOCIATION, CODE_MISSING_USER_TENANT_ASSOCIATION))));
+              promise.fail(new Exception(MISSING_USER_TENANT_ASSOCIATION));
+            } else {
+              promise.complete(tenantId);
+            }
+          } else if (userTenants.size() > 1) {
+            List<String> matchingTenants = userTenants.stream()
+                .map(UserTenant::getTenantId)
+                .collect(toList());
+            logger.debug(MULTIPLE_MATCHING_USERS_LOG, credentials.getUsername(), credentials.getUserId(),
+                matchingTenants.toString());
+            asyncResultHandler.handle(Future.succeededFuture(PostAuthnLoginResponse.respond422WithApplicationJson(
```

Comment:
It looks like a list of tenant Ids are returned as a response. Is the UI supposed to get tenant names from mod-consortia?

Reply:
"Tenant ids" (or simply "tenants") are not UUIDs, they are names like "diku". I don't know if there are more readable names associated with them. If there are, then I guess the UI would need them.